### PR TITLE
rbac: improve 'create-for-rbac'

### LINKED ISSUE #1190: "az ad sp create-for-rbac" should set sensible defaults
"az ad sp create-for-rbac" does not set sensible defaults and the resulting principal is not very usable.  It would be great if it would setup defaults as the following "az ad sp create-for-rbac --role contributor --scopes /subscriptions/SUBSCRIPTIONID".

### LINKED ISSUE #1332: az ad sp create-for-rbac: flakiness
The following cmd sometimes returns error, but the actual AAD app is created (but not the SP):

```
$ az ad sp create-for-rbac --scopes /subscriptions/xxxxxx/resourcegroups/gggggg/providers/Microsoft.ContainerRegistry/registries/nnnn \
 --role Owner --password "$azure_client_secret" \
 --years 2 --name http://foo-bar-ahmetb/

ERROR: Resource 'ServicePrincipal_00e1c5b1-394e-48ca-a1b7-f39b77635928' does not exist or one of its queried reference-property objects are not present.
```

When I do `az ad sp show --id http://foo-bar-ahmetb/`, I got error:

    ERROR: service principal 'http://foo-bar-ahmetb/' doesn't exist

but `az ad app show --id http://foo-bar-ahmetb/` worked just fine. 

So it looks like `create-for-rbac` probably is running into some replication delay issue or somethign else from AAD and therefore erroring out, possibly having created the AAD App but not the SP.

### PR DESCRIPTION
**Pull request by yugangw-msft**
Fix #1190, Fix #1332
Includes:
1. Make it reliable with retires, while the root cause of the replicate latency is being tracked by ICM ticket and investigated by AAD service team  
2. Default to create a role assignment, using the scope of current subscription root with role of "contributor".
3. Support re-entry in the command. Say, if you have SP created in the previous run, you can run it again to continue to the end

//cc: @colemickens, @ahmetalpbalkan

### CODE REVIEW

**Comment by yugangw-msft** [comment]
//cc: @derekbekoe @tjprescott

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -93,8 +93,8 @@ def _search_role_definitions(definitions_client, name, scope, custom_role_only=F
 def create_role_assignment(role, assignee, resource_group_name=None, scope=None):
     return _create_role_assignment(role, assignee, resource_group_name, scope)
 
-def _create_role_assignment(role, assignee, resource_group_name=None, scope=None,
-                            ocp_aad_session_key=None):
+def _create_role_assignment(role, assignee, resource_group_name=None, scope=None, #pylint: disable=too-many-arguments
+                            ocp_aad_session_key=None, resolve_assignee=True):
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Should we have help for this parameter?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -399,18 +399,22 @@ def _build_application_creds(password=None, key_value=None, key_type=None,#pylin
 def create_service_principal(identifier):
     return _create_service_principal(identifier)
 
-def _create_service_principal(identifier, retain_raw_response=False):
+def _create_service_principal(identifier, retain_raw_response=False, resolve_app=True):
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Add help for this parameter?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 3]
```diff
@@ -434,55 +438,98 @@ def _resolve_service_principal(client, identifier):
...
-    oid = aad_sp.output.object_id if scopes else aad_sp.object_id
-
-    if scopes:
-        #It is possible the SP has not been propagated to all servers, so creating assignments
-        #might fail. The reliable workaround is to call out the server where creation occurred.
-        #pylint: disable=protected-access
+    import time
+    graph_client = _graph_client_factory()
+    role_client = _auth_client_factory().role_assignments
+    role = role or 'contributor'
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: The parameter default is `Contributor`. Does the casing matter?
Actually, do you need this or statement if there is a default set at the top?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 4]
```diff
@@ -434,55 +438,98 @@ def _resolve_service_principal(client, identifier):
...
+            try:
+                aad_sp = _create_service_principal(app_id, bool(scopes), resolve_app=False)
+                break
+            except Exception as ex: #pylint: disable=broad-except
+                #pylint: disable=line-too-long
+                if 'The appId of the service principal does not reference a valid application object' in str(ex):
+                    time.sleep(5)
+                else:
+                    logger.warning("Creating service principal failed for appid '%s'. Trace followed:\n%s",
+                                   name, ex.response.headers) #pylint: disable=no-member
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Should we log the detailed trace to logger.debug and then `raise CLIError("Creating service principal failed for appid '%s'.")`.

It appears that logging a warning and then calling `raise` would print the full stacktrace to the console?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 5]
```diff
@@ -434,55 +438,98 @@ def _resolve_service_principal(client, identifier):
...
+        #retry while the root cause is being investigated by AAD service team
+        #And the 'session_key' will not be used as well.
+        for _ in range(1, 12):
+            try:
+                aad_sp = _create_service_principal(app_id, bool(scopes), resolve_app=False)
+                break
+            except Exception as ex: #pylint: disable=broad-except
+                #pylint: disable=line-too-long
+                if 'The appId of the service principal does not reference a valid application object' in str(ex):
+                    time.sleep(5)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Consider: Add logger.warning message indicating that it is retrying something and what retry number you're currently on e.g. (Retry 3/12).

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 6]
```diff
@@ -434,55 +438,98 @@ def _resolve_service_principal(client, identifier):
...
+
+    #Again, use retry while replicate latency is being investigated
+    for scope in scopes:
+        for _ in range(1, 24):
+            try:
+                _create_role_assignment(role, sp_oid, None, scope, resolve_assignee=False)
+                break
+            except Exception as ex:
+                if ' does not exist in the directory ' in str(ex):
+                    time.sleep(5)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Again, suggest retry warning messages so user is informed.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 1 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: this is an internal method which is not exposed as a command

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 2 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Same like my previous comment

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 3 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Casing doesn't matter, but yesm this code is redundant as we have a default set at the top

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 4 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Let us do it later.
For now, logging as a warn by default will display the correlation id to give it to the service team to investigate. Using `--debug` changes the timing and hence it becomes hard to reproduce.

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 5 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: will do

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 6 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: will do

**Comment by yugangw-msft** [comment]
@derekbekoe, please take another look.

**Comment by derekbekoe** [comment]
@yugangw-msft Have the changes been pushed here?

**Comment by yugangw-msft** [comment]
@derekbekoe, sorry, I just now pushed.

**Pull request merged**